rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+*   Trainer that wraps standard TF.Estimator or Keras code
+*   Evaluator to generate, slice, and visualize evaluation metrics
+*   Custom analyses on the output of any of these components with standard
+    Python
+
+To close the loop, the notebook will be automatically exported as a pipeline
+configuration that users can directly deploy as a scalable TFX pipeline. There
+is no additional modification required.
+
+## Target Users
+
+We target users who want to manually iterate on their models & components, and
+prefer a notebook environment for the benefits outlined above. This is a wide
+range of potential users, and from our user research, spans software engineers
+and data scientists within and outside of Google.
+
+## Design Proposal
+
+This proposal proposes a set of primitives that match concepts in the current
+TFX SDK.
+
+### Component definition; inputs and outputs
+
+#### Proposal: components should take inputs, produce outputs (instead of taking predefined upstream components)
+
+This proposal proposes a set of primitives that match concepts in the current
+TFX SDK. We propose to follow the current TFX style of having components
+explicitly take input channels (i.e. streams of artifacts of a specific type)
+and produce output channels (of another specific type). This could look like
+this:
+
+```
+# Here, with an input_base as an execution parameter with a given
+# file path.
+example_gen = CsvExampleGen(input_base=examples)
+
+# Next, we use the 'examples' named output of ExampleGen as the
+# input to StatisticsGen.
+statistics_gen = StatisticsGen(input_data=example_gen.outputs['examples'])
+
+# We then similarly use the statsgen output in SchemaGen.
+infer_schema = SchemaGen(stats=statistics_gen.outputs['output'])
+
+# Next, we do example validation.
+validate_stats = ExampleValidator(
+    stats=statistics_gen.outputs['output'],
+    schema=infer_schema.outputs['output'])
+```
+
+### Component execution, execution result objects, visualization
+
+#### Proposal: InteractiveContext.run(component) returns an ExecutionResult, whose output artifacts can be visualized using InteractiveContext.show(artifacts)
+
+##### Part 1 (Option 1): add InteractiveContext.run() **[recommended]**
+
+We propose to add a new `InteractiveContext` class. Distinct from a pipeline
+runner which takes in an entire TFX pipeline, an instance of this class allows
+interactive execution of individual components. Here, a user would construct
+components with appropriate parameters and execution properties, and the
+`InteractiveContext.run(component)` method would execute that component, thereby
+materializing any output artifacts of that component.
+
+An advantage of this style is that it does not bifurcate the TFX pipeline runner
+concept into "pipeline runners" and "component runners", and it is very clear
+that this API is only meant for interactive usage (as opposed to the two
+alternatives below). A disadvantage is that we may not want to introduce
```

Comment:
What sort of backward compatibility guarantees are you intending to offer on all of these APIs?

Reply:
With the existing TFX API as a baseline, the component DAG configuration aspects are not modified in this proposal, so there are no additional compatibility guarantees.  The only additional API we propose to add here would be the run() functionality, for which the guarantee is that in a notebook, it will continue to provide the functionality of running a specified component.